File: android/src/main/java/expo/modules/mpvplayer/MPVRenderer.kt
package expo.modules.mpvplayer

import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.view.Surface
import java.io.File

data class MpvLoadConfig(
  val url: String,
  val headers: Map<String, String> = emptyMap(),
  val externalSubtitles: List<String> = emptyList(),
  val startPosition: Double? = null,
  val autoplay: Boolean = true,
  val initialAudioId: Int? = null,
  val initialSubtitleId: Int? = null,
  val cacheEnabled: String? = null,
  val cacheSeconds: Int? = null,
  val maxBytes: Int? = null,
  val maxBackBytes: Int? = null,
  val voDriver: String = "gpu-next",
)

data class MpvTrack(
  val id: Int,
  val type: String,
  val title: String?,
  val lang: String?,
  val codec: String?,
  val channels: Int?,
  val selected: Boolean,
)

data class MpvTechnicalInfo(
  val videoWidth: Int?,
  val videoHeight: Int?,
  val videoCodec: String?,
  val audioCodec: String?,
  val fps: Double?,
  val videoBitrate: Int?,
  val audioBitrate: Int?,
  val cacheSeconds: Double?,
  val droppedFrames: Int?,
  val voDriver: String?,
  val hwdec: String?,
)

interface MpvRendererDelegate {
  fun onLoad(url: String)
  fun onProgress(position: Double, duration: Double, cacheSeconds: Double)
  fun onPlaybackState(
    isPaused: Boolean?,
    isPlaying: Boolean?,
    isLoading: Boolean?,
    isReadyToSeek: Boolean?,
  )

  fun onTracksReady()
  fun onError(message: String)
}

/**
 * High-level libmpv driver for Android. Configures the VO for the LGPL Android
 * GL path, loads with HTTP headers, observes properties, throttles `time-pos`
 * to ~1/sec (bypassed while seeking), enumerates tracks, and surfaces
 * diagnostics. All delegate callbacks are posted to the main thread.
 */
class MPVRenderer(context: Context, delegate: MpvRendererDelegate) : MpvLib.Listener {
  private val mpv = MpvLib(context)
  private val main = Handler(Looper.getMainLooper())
  private var delegate: MpvRendererDelegate? = delegate

  private var pendingConfig: MpvLoadConfig? = null
  private var lastProgressEmit = 0L
  private var seeking = false

  private companion object {
    const val PROGRESS_INTERVAL_MS = 1000L
    const val SUBFONT_ASSET = "subfont.ttf"
  }

  init {
    mpv.listener = this
    val configDir = prepareConfigDir(context)
    mpv.setOption("config", "yes")
    mpv.setOption("config-dir", configDir)
    mpv.setOption("gpu-context", "android")
    mpv.setOption("opengl-es", "yes")
    mpv.setOption("hwdec", if (isEmulator()) "no" else "mediacodec-copy")
    mpv.setOption("hwdec-codecs", "h264,hevc,mpeg4,mpeg2video,vp8,vp9,av1")
    mpv.setOption("tls-verify", "no")
    mpv.setOption("cache", "yes")
    mpv.setOption("demuxer-max-bytes", "64MiB")
    mpv.setOption("force-window", "no")
    mpv.setOption("keep-open", "always")
    mpv.setOption("ytdl", "no")
  }

  fun initialize(voDriver: String) {
    mpv.setOption("vo", voDriver)
    mpv.initialize()
    observeProperties()
  }

  private fun observeProperties() {
    mpv.observe("time-pos", MpvFormat.DOUBLE)
    mpv.observe("duration", MpvFormat.DOUBLE)
    mpv.observe("pause", MpvFormat.FLAG)
    mpv.observe("paused-for-cache", MpvFormat.FLAG)
    mpv.observe("demuxer-cache-duration", MpvFormat.DOUBLE)
    mpv.observe("track-list/count", MpvFormat.INT64)
    mpv.observe("eof-reached", MpvFormat.FLAG)
  }

  // MARK: - Surface

  fun attachSurface(surface: Surface) {
    mpv.attachSurface(surface)
    mpv.setOption("force-window", "yes")
  }

  fun setSurfaceSize(width: Int, height: Int) = mpv.setSurfaceSize(width, height)

  fun detachSurface() {
    // Keep the VO alive (do NOT set vo=null) so re-attach doesn't black-screen.
    mpv.detachSurface()
  }

  // MARK: - Load & transport

  fun load(config: MpvLoadConfig) {
    if (config.headers.isNotEmpty()) {
      val joined = config.headers.entries.joinToString(",") { "${it.key}: ${it.value}" }
      mpv.setOption("http-header-fields", joined)
    }
    config.cacheEnabled?.let { mpv.setOption("cache", it) }
    config.cacheSeconds?.let { mpv.setOption("cache-secs", it.toString()) }
    config.maxBytes?.let { mpv.setOption("demuxer-max-bytes", it.toString()) }
    config.maxBackBytes?.let { mpv.setOption("demuxer-max-back-bytes", it.toString()) }
    config.startPosition?.takeIf { it > 0 }?.let { mpv.setOption("start", it.toString()) }

    mpv.setBoolean("pause", !config.autoplay)
    pendingConfig = config
    delegate?.let { d -> main.post { d.onPlaybackState(null, null, true, false) } }
    mpv.command(arrayOf("loadfile", config.url, "replace"))
  }

  fun play() = mpv.setBoolean("pause", false)

  fun pause() = mpv.setBoolean("pause", true)

  fun seekTo(seconds: Double) {
    seeking = true
    mpv.command(arrayOf("seek", seconds.toString(), "absolute"))
  }

  fun seekBy(seconds: Double) {
    seeking = true
    mpv.command(arrayOf("seek", seconds.toString(), "relative"))
  }

  fun setSpeed(rate: Double) = mpv.setDouble("speed", rate)

  fun getSpeed(): Double = mpv.getDouble("speed") ?: 1.0

  fun isPaused(): Boolean = mpv.getBoolean("pause") ?: true

  fun getCurrentPosition(): Double = mpv.getDouble("time-pos") ?: 0.0

  fun getDuration(): Double = mpv.getDouble("duration") ?: 0.0

  // MARK: - Tracks

  fun tracks(type: String): List<MpvTrack> {
    val count = mpv.getInt("track-list/count") ?: return emptyList()
    val result = mutableListOf<MpvTrack>()
    for (i in 0 until count) {
      if (mpv.getString("track-list/$i/type") != type) continue
      val id = mpv.getInt("track-list/$i/id") ?: continue
      result.add(
        MpvTrack(
          id = id,
          type = type,
          title = mpv.getString("track-list/$i/title"),
          lang = mpv.getString("track-list/$i/lang"),
          codec = mpv.getString("track-list/$i/codec"),
          channels = mpv.getInt("track-list/$i/audio-channels"),
          selected = mpv.getBoolean("track-list/$i/selected") ?: false,
        )
      )
    }
    return result
  }

  fun setAudioTrack(id: Int) = mpv.setString("aid", id.toString())

  fun setSubtitleTrack(id: Int) = mpv.setString("sid", id.toString())

  fun disableSubtitles() = mpv.setString("sid", "no")

  fun currentTrackId(property: String): Int? {
    if (mpv.getString(property) == "no") return null
    return mpv.getInt(property)
  }

  fun addSubtitleFile(url: String, select: Boolean) {
    mpv.command(arrayOf("sub-add", url, if (select) "select" else "auto"))
  }

  // MARK: - Video scaling

  fun setZoomedToFill(zoom: Boolean) = mpv.setString("panscan", if (zoom) "1.0" else "0.0")

  fun isZoomedToFill(): Boolean = (mpv.getDouble("panscan") ?: 0.0) > 0.5

  // MARK: - Subtitle styling & A/V sync

  fun setSubtitleScale(scale: Double) = mpv.setDouble("sub-scale", scale)

  fun setSubtitlePosition(position: Double) = mpv.setDouble("sub-pos", position)

  fun setSubtitleDelay(seconds: Double) = mpv.setDouble("sub-delay", seconds)

  fun setAudioDelay(seconds: Double) = mpv.setDouble("audio-delay", seconds)

  // MARK: - Diagnostics

  fun technicalInfo(): MpvTechnicalInfo =
    MpvTechnicalInfo(
      videoWidth = mpv.getInt("width"),
      videoHeight = mpv.getInt("height"),
      videoCodec = mpv.getString("video-codec"),
      audioCodec = mpv.getString("audio-codec"),
      fps = mpv.getDouble("container-fps") ?: mpv.getDouble("estimated-vf-fps"),
      videoBitrate = mpv.getInt("video-bitrate"),
      audioBitrate = mpv.getInt("audio-bitrate"),
      cacheSeconds = mpv.getDouble("demuxer-cache-duration"),
      droppedFrames = mpv.getInt("frame-drop-count"),
      voDriver = mpv.getString("current-vo"),
      hwdec = mpv.getString("hwdec-current"),
    )

  fun destroy() {
    delegate = null
    mpv.destroy()
  }

  // MARK: - MpvLib.Listener (libmpv thread -> main thread)

  override fun onEvent(eventId: Int) {
    when (eventId) {
      MpvEvent.FILE_LOADED -> {
        applyPendingSelections()
        val url = pendingConfig?.url
        main.post {
          url?.let { delegate?.onLoad(it) }
          delegate?.onPlaybackState(null, null, false, true)
          delegate?.onTracksReady()
        }
      }
      MpvEvent.PLAYBACK_RESTART -> {
        seeking = false
        main.post { delegate?.onPlaybackState(null, null, false, true) }
      }
      MpvEvent.END_FILE -> main.post { delegate?.onPlaybackState(null, false, false, false) }
      MpvEvent.SHUTDOWN -> {}
    }
  }

  override fun onProperty(name: String) {}

  override fun onProperty(name: String, value: Long) {}

  override fun onProperty(name: String, value: Double) {
    if (name == "time-pos") emitProgressIfDue()
  }

  override fun onProperty(name: String, value: Boolean) {
    when (name) {
      "pause" -> main.post { delegate?.onPlaybackState(value, !value, null, null) }
      "paused-for-cache" -> main.post { delegate?.onPlaybackState(null, null, value, null) }
    }
  }

  override fun onProperty(name: String, value: String) {}

  private fun emitProgressIfDue() {
    val now = System.currentTimeMillis()
    if (!seeking && now - lastProgressEmit < PROGRESS_INTERVAL_MS) return
    lastProgressEmit = now
    val position = mpv.getDouble("time-pos") ?: 0.0
    val duration = mpv.getDouble("duration") ?: 0.0
    val cache = mpv.getDouble("demuxer-cache-duration") ?: 0.0
    main.post { delegate?.onProgress(position, duration, cache) }
  }

  private fun applyPendingSelections() {
    val config = pendingConfig ?: return
    config.initialAudioId?.let { setAudioTrack(it) }
    config.initialSubtitleId?.let { setSubtitleTrack(it) }
    config.externalSubtitles.forEach { mpv.command(arrayOf("sub-add", it, "auto")) }
  }

  // MARK: - Setup helpers

  /** Copies the bundled subtitle font into mpv's config dir (SRT needs it). */
  private fun prepareConfigDir(context: Context): String {
    val dir = File(context.filesDir, "mpv")
    if (!dir.exists()) dir.mkdirs()
    val font = File(dir, SUBFONT_ASSET)
    if (!font.exists()) {
      runCatching {
        context.assets.open(SUBFONT_ASSET).use { input ->
          font.outputStream().use { output -> input.copyTo(output) }
        }
      }
    }
    return dir.path
  }

  private fun isEmulator(): Boolean {
    val fp = Build.FINGERPRINT
    return fp.startsWith("generic") ||
      fp.contains("emulator", ignoreCase = true) ||
      Build.HARDWARE.contains("ranchu") ||
      Build.HARDWARE.contains("goldfish") ||
      Build.MODEL.contains("sdk", ignoreCase = true)
  }
}

File: ios/MPVRenderer.swift
import Foundation
import Libmpv

/// Snapshot of low-level playback diagnostics (mirrors the JS `TechnicalInfo`).
struct MPVTechnicalInfo {
  var videoWidth: Int?
  var videoHeight: Int?
  var videoCodec: String?
  var audioCodec: String?
  var fps: Double?
  var videoBitrate: Int?
  var audioBitrate: Int?
  var cacheSeconds: Double?
  var droppedFrames: Int?
  var voDriver: String?
  var hwdec: String?
}

struct MPVTrack {
  var id: Int
  var type: String
  var title: String?
  var lang: String?
  var codec: String?
  var channels: Int?
  var selected: Bool
}

/// Options used to (re)load a single file. Built from the JS `source` prop.
struct MPVLoadConfig {
  var url: String
  var headers: [String: String] = [:]
  var externalSubtitles: [String] = []
  var startPosition: Double?
  var autoplay: Bool = true
  var initialAudioId: Int?
  var initialSubtitleId: Int?
  var cacheEnabled: String?
  var cacheSeconds: Int?
  var maxBytes: Int?
  var maxBackBytes: Int?
  var allowSelfSignedTls: Bool = false
}

protocol MPVRendererDelegate: AnyObject {
  func rendererDidLoad(url: String)
  func rendererDidUpdateProgress(position: Double, duration: Double, cacheSeconds: Double)
  func rendererDidChangePlaybackState(
    isPaused: Bool?, isPlaying: Bool?, isLoading: Bool?, isReadyToSeek: Bool?)
  func rendererTracksReady()
  func rendererDidError(_ message: String)
}

/// Owns the raw libmpv handle and drives the canonical LGPL Metal path
/// (`vo=gpu-next` + `gpu-api=vulkan` + `gpu-context=moltenvk`). All event
/// draining happens on a background queue; the delegate is responsible for
/// hopping to the main thread before touching JS/UIKit.
final class MPVRenderer {
  weak var delegate: MPVRendererDelegate?

  private var mpv: OpaquePointer?
  private let eventQueue = DispatchQueue(label: "expo-mpv-player.events", qos: .userInitiated)

  // ~1/sec throttle for time-pos driven progress; bypassed while seeking.
  private var lastProgressEmit: TimeInterval = 0
  private var isSeeking = false
  private let progressInterval: TimeInterval = 1.0

  init(layer: MPVMetalLayer, isSimulator: Bool) {
    mpv = mpv_create()
    guard let mpv else {
      delegate?.rendererDidError("mpv_create() returned null")
      return
    }

    // Attach the CAMetalLayer as the window id BEFORE initialize.
    var layerPtr = Int64(Int(bitPattern: Unmanaged.passUnretained(layer).toOpaque()))
    mpv_set_option(mpv, "wid", MPV_FORMAT_INT64, &layerPtr)

    setOption("vo", "gpu-next")
    setOption("gpu-api", "vulkan")
    setOption("gpu-context", "moltenvk")
    setOption("hwdec", isSimulator ? "no" : "videotoolbox")
    setOption("hwdec-codecs", "all")
    // tls-verify is set per-source in load() — secure (validating) by default.
    setOption("ytdl", "no")
    setOption("keep-open", "always")
    setOption("cache", "yes")

    mpv_request_log_messages(mpv, "warn")

    let status = mpv_initialize(mpv)
    if status < 0 {
      delegate?.rendererDidError("mpv_initialize failed: \(errorString(status))")
      return
    }

    observeProperties()

    // Drain events off the main thread whenever mpv wakes us.
    mpv_set_wakeup_callback(
      mpv,
      { ctx in
        guard let ctx else { return }
        let renderer = Unmanaged<MPVRenderer>.fromOpaque(ctx).takeUnretainedValue()
        renderer.eventQueue.async { renderer.drainEvents() }
      }, Unmanaged.passUnretained(self).toOpaque())
  }

  private var destroyed = false

  deinit {
    invalidate()
  }

  /// Tears down mpv safely. Serialized on the event queue so it can't race with
  /// drainEvents(), and idempotent so an explicit teardown and MPV_EVENT_SHUTDOWN
  /// can't double-destroy. Call this from the view's teardown rather than relying
  /// on `deinit` timing.
  func invalidate() {
    eventQueue.sync { destroyMpv() }
  }

  private func destroyMpv() {
    guard !destroyed, let mpv else { return }
    destroyed = true
    mpv_set_wakeup_callback(mpv, nil, nil)
    mpv_terminate_destroy(mpv)
    self.mpv = nil
  }

  // MARK: - Load & transport

  func load(_ config: MPVLoadConfig) {
    guard let mpv else { return }

    // HTTP auth headers: set BEFORE loadfile via `change-list … append` so each
    // header is a single list item. A value containing a comma (e.g. a Jellyfin
    // `MediaBrowser Client="x", Token="y"`) is NOT split into bogus headers the
    // way a naive comma-join would. Clear first so a reload doesn't accumulate.
    command(["change-list", "http-header-fields", "clr", ""])
    for (key, value) in config.headers {
      command(["change-list", "http-header-fields", "append", "\(key): \(value)"])
    }

    // After mpv_initialize, options must be set via the property API (not
    // mpv_set_option_string, which is for pre-init configuration).
    if let enabled = config.cacheEnabled { setPropertyString("cache", value: enabled) }
    if let secs = config.cacheSeconds { setPropertyString("cache-secs", value: String(secs)) }
    if let bytes = config.maxBytes { setPropertyString("demuxer-max-bytes", value: String(bytes)) }
    if let back = config.maxBackBytes {
      setPropertyString("demuxer-max-back-bytes", value: String(back))
    }

    // Resume position (applied on the next load).
    if let start = config.startPosition, start > 0 {
      setPropertyString("start", value: String(start))
    }

    // TLS: validate certs by default; only disable for an opted-in source.
    setPropertyString("tls-verify", value: config.allowSelfSignedTls ? "no" : "yes")

    // Honour autoplay: start paused if not autoplaying.
    setPropertyFlag("pause", value: !config.autoplay)

    pendingConfig = config
    delegate?.rendererDidChangePlaybackState(
      isPaused: nil, isPlaying: nil, isLoading: true, isReadyToSeek: false)
    command(["loadfile", config.url, "replace"])
  }

  private var pendingConfig: MPVLoadConfig?

  func play() { setPropertyFlag("pause", value: false) }
  func pause() { setPropertyFlag("pause", value: true) }

  func seek(to seconds: Double) {
    isSeeking = true
    command(["seek", String(seconds), "absolute"])
  }

  func seekBy(_ seconds: Double) {
    isSeeking = true
    command(["seek", String(seconds), "relative"])
  }

  func setSpeed(_ rate: Double) {
    var value = rate
    if let mpv { mpv_set_property(mpv, "speed", MPV_FORMAT_DOUBLE, &value) }
  }

  // MARK: - Queries

  func getSpeed() -> Double { getDouble("speed") ?? 1.0 }
  func isPaused() -> Bool { getFlag("pause") ?? true }
  func getCurrentPosition() -> Double { getDouble("time-pos") ?? 0 }
  func getDuration() -> Double { getDouble("duration") ?? 0 }

  // MARK: - Tracks

  func tracks(ofType type: String) -> [MPVTrack] {
    guard let count = getInt("track-list/count") else { return [] }
    var result: [MPVTrack] = []
    for i in 0..<count {
      guard getString("track-list/\(i)/type") == type else { continue }
      guard let id = getInt("track-list/\(i)/id") else { continue }
      result.append(
        MPVTrack(
          id: id,
          type: type,
          title: getString("track-list/\(i)/title"),
          lang: getString("track-list/\(i)/lang"),
          codec: getString("track-list/\(i)/codec"),
          channels: getInt("track-list/\(i)/audio-channels"),
          selected: getFlag("track-list/\(i)/selected") ?? false
        ))
    }
    return result
  }

  func setAudioTrack(_ id: Int) { setPropertyString("aid", value: String(id)) }
  func setSubtitleTrack(_ id: Int) { setPropertyString("sid", value: String(id)) }
  func disableSubtitles() { setPropertyString("sid", value: "no") }

  func currentTrackId(_ property: String) -> Int? {
    // mpv returns "no" (string) when disabled.
    if getString(property) == "no" { return nil }
    return getInt(property)
  }

  func addSubtitleFile(_ url: String, select: Bool) {
    command(["sub-add", url, select ? "select" : "auto"])
  }

  // MARK: - Video scaling

  func setZoomedToFill(_ zoom: Bool) {
    // panscan: 0.0 fit (letterbox), 1.0 fill (crop).
    setPropertyString("panscan", value: zoom ? "1.0" : "0.0")
  }

  func isZoomedToFill() -> Bool { (getDouble("panscan") ?? 0) > 0.5 }

  // MARK: - Subtitle styling & A/V sync

  func setSubtitleScale(_ scale: Double) { setPropertyDouble("sub-scale", scale) }
  func setSubtitlePosition(_ position: Double) { setPropertyDouble("sub-pos", position) }
  func setSubtitleDelay(_ seconds: Double) { setPropertyDouble("sub-delay", seconds) }
  func setAudioDelay(_ seconds: Double) { setPropertyDouble("audio-delay", seconds) }

  // MARK: - Diagnostics

  func technicalInfo() -> MPVTechnicalInfo {
    MPVTechnicalInfo(
      videoWidth: getInt("width"),
      videoHeight: getInt("height"),
      videoCodec: getString("video-codec"),
      audioCodec: getString("audio-codec"),
      fps: getDouble("container-fps") ?? getDouble("estimated-vf-fps"),
      videoBitrate: getInt("video-bitrate"),
      audioBitrate: getInt("audio-bitrate"),
      cacheSeconds: getDouble("demuxer-cache-duration"),
      droppedFrames: getInt("frame-drop-count"),
      voDriver: getString("current-vo"),
      hwdec: getString("hwdec-current")
    )
  }

  // MARK: - Property observation

  private func observeProperties() {
    guard let mpv else { return }
    mpv_observe_property(mpv, 0, "time-pos", MPV_FORMAT_DOUBLE)
    mpv_observe_property(mpv, 0, "duration", MPV_FORMAT_DOUBLE)
    mpv_observe_property(mpv, 0, "pause", MPV_FORMAT_FLAG)
    mpv_observe_property(mpv, 0, "paused-for-cache", MPV_FORMAT_FLAG)
    mpv_observe_property(mpv, 0, "demuxer-cache-duration", MPV_FORMAT_DOUBLE)
    mpv_observe_property(mpv, 0, "eof-reached", MPV_FORMAT_FLAG)
  }

  // MARK: - Event loop

  private func drainEvents() {
    guard let mpv else { return }
    while true {
      guard let eventPtr = mpv_wait_event(mpv, 0) else { break }
      let event = eventPtr.pointee
      if event.event_id == MPV_EVENT_NONE { break }
      handle(event)
    }
  }

  private func handle(_ event: mpv_event) {
    switch event.event_id {
    case MPV_EVENT_FILE_LOADED:
      applyPendingSelections()
      if let url = pendingConfig?.url { delegate?.rendererDidLoad(url: url) }
      delegate?.rendererDidChangePlaybackState(
        isPaused: nil, isPlaying: nil, isLoading: false, isReadyToSeek: true)
      delegate?.rendererTracksReady()

    case MPV_EVENT_PLAYBACK_RESTART:
      isSeeking = false
      delegate?.rendererDidChangePlaybackState(
        isPaused: nil, isPlaying: nil, isLoading: false, isReadyToSeek: true)

    case MPV_EVENT_END_FILE:
      // Surface load/decode failures (401, bad URL, unsupported, …) as onError.
      // A clean EOF/stop has a non-ERROR reason and must not fire onError.
      if let endFile = event.data?.assumingMemoryBound(to: mpv_event_end_file.self).pointee,
        endFile.reason == MPV_END_FILE_REASON_ERROR
      {
        delegate?.rendererDidError(String(cString: mpv_error_string(endFile.error)))
      }
      delegate?.rendererDidChangePlaybackState(
        isPaused: nil, isPlaying: false, isLoading: false, isReadyToSeek: false)

    case MPV_EVENT_PROPERTY_CHANGE:
      handlePropertyChange(event)

    case MPV_EVENT_SHUTDOWN:
      destroyMpv()

    case MPV_EVENT_LOG_MESSAGE:
      if let data = event.data?.assumingMemoryBound(to: mpv_event_log_message.self).pointee,
        let text = data.text
      {
        MpvLog.warn(String(cString: text).trimmingCharacters(in: .whitespacesAndNewlines))
      }

    default:
      break
    }
  }

  private func handlePropertyChange(_ event: mpv_event) {
    guard let data = event.data?.assumingMemoryBound(to: mpv_event_property.self).pointee,
      let namePtr = data.name
    else { return }
    let name = String(cString: namePtr)

    switch name {
    case "time-pos":
      emitProgressIfDue()
    case "pause":
      if data.format == MPV_FORMAT_FLAG, let flag = data.data?.assumingMemoryBound(to: Int32.self).pointee {
        let paused = flag != 0
        delegate?.rendererDidChangePlaybackState(
          isPaused: paused, isPlaying: !paused, isLoading: nil, isReadyToSeek: nil)
      }
    case "paused-for-cache":
      if data.format == MPV_FORMAT_FLAG, let flag = data.data?.assumingMemoryBound(to: Int32.self).pointee {
        delegate?.rendererDidChangePlaybackState(
          isPaused: nil, isPlaying: nil, isLoading: flag != 0, isReadyToSeek: nil)
      }
    default:
      break
    }
  }

  private func emitProgressIfDue() {
    let now = Date().timeIntervalSince1970
    if !isSeeking && now - lastProgressEmit < progressInterval { return }
    lastProgressEmit = now
    let position = getDouble("time-pos") ?? 0
    let duration = getDouble("duration") ?? 0
    let cache = getDouble("demuxer-cache-duration") ?? 0
    delegate?.rendererDidUpdateProgress(
      position: position, duration: duration, cacheSeconds: cache)
  }

  private func applyPendingSelections() {
    guard let config = pendingConfig else { return }
    if let aid = config.initialAudioId { setAudioTrack(aid) }
    if let sid = config.initialSubtitleId { setSubtitleTrack(sid) }
    for sub in config.externalSubtitles {
      command(["sub-add", sub, "auto"])
    }
  }

  // MARK: - libmpv helpers

  private func setOption(_ name: String, _ value: String) {
    guard let mpv else { return }
    mpv_set_option_string(mpv, name, value)
  }

  private func setPropertyString(_ name: String, value: String) {
    guard let mpv else { return }
    mpv_set_property_string(mpv, name, value)
  }

  private func setPropertyFlag(_ name: String, value: Bool) {
    guard let mpv else { return }
    var flag: Int32 = value ? 1 : 0
    mpv_set_property(mpv, name, MPV_FORMAT_FLAG, &flag)
  }

  private func setPropertyDouble(_ name: String, _ value: Double) {
    guard let mpv else { return }
    var v = value
    mpv_set_property(mpv, name, MPV_FORMAT_DOUBLE, &v)
  }

  private func getDouble(_ name: String) -> Double? {
    guard let mpv else { return nil }
    var out: Double = 0
    return mpv_get_property(mpv, name, MPV_FORMAT_DOUBLE, &out) >= 0 ? out : nil
  }

  private func getInt(_ name: String) -> Int? {
    guard let mpv else { return nil }
    var out: Int64 = 0
    return mpv_get_property(mpv, name, MPV_FORMAT_INT64, &out) >= 0 ? Int(out) : nil
  }

  private func getFlag(_ name: String) -> Bool? {
    guard let mpv else { return nil }
    var out: Int32 = 0
    return mpv_get_property(mpv, name, MPV_FORMAT_FLAG, &out) >= 0 ? out != 0 : nil
  }

  private func getString(_ name: String) -> String? {
    guard let mpv else { return nil }
    guard let cstr = mpv_get_property_string(mpv, name) else { return nil }
    defer { mpv_free(cstr) }
    let value = String(cString: cstr)
    return value.isEmpty ? nil : value
  }

  private func command(_ args: [String]) {
    guard let mpv else { return }
    // mpv_command wants a NULL-terminated `const char *[]`.
    var cargs: [UnsafePointer<CChar>?] = args.map { UnsafePointer(strdup($0)) }
    cargs.append(nil)
    defer {
      for ptr in cargs where ptr != nil {
        free(UnsafeMutablePointer(mutating: ptr))
      }
    }
    cargs.withUnsafeMutableBufferPointer { buf in
      _ = mpv_command(mpv, buf.baseAddress)
    }
  }

  private func errorString(_ status: Int32) -> String {
    String(cString: mpv_error_string(status))
  }
}
